sdk/storage/src/blobs/blob_client.cpp

```diff
@@ -0,0 +1,179 @@
+// SPDX-License-Identifier: MIT
+
+#include "blobs/blob_client.hpp"
+
+#include "common/storage_common.hpp"
+#include "http/curl/curl.hpp"
+
+namespace Azure { namespace Storage { namespace Blobs {
+
+  BlobClient BlobClient::FromConnectionString(
+      const std::string& connectionString,
+      const std::string& containerName,
+      const std::string& blobName,
+      const BlobClientOptions& options)
+  {
+    auto parsedConnectionString = ParseConnectionString(connectionString);
+
+    std::string accountName;
+    std::string accountKey;
+    std::string blobEndpoint;
+    std::string EndpointSuffix;
+    std::string defaultEndpointsProtocol = ".core.windows.net";
+
+    auto ite = parsedConnectionString.find("AccountName");
+    if (ite != parsedConnectionString.end())
+    {
+      accountName = ite->second;
+    }
+    ite = parsedConnectionString.find("AccountKey");
+    if (ite != parsedConnectionString.end())
+    {
+      accountKey = ite->second;
+    }
+    ite = parsedConnectionString.find("BlobEndpoint");
+    if (ite != parsedConnectionString.end())
+    {
+      blobEndpoint = ite->second;
+    }
+    ite = parsedConnectionString.find("EndpointSuffix");
+    if (ite != parsedConnectionString.end())
+    {
+      EndpointSuffix = ite->second;
+    }
+    ite = parsedConnectionString.find("DefaultEndpointsProtocol");
+    if (ite != parsedConnectionString.end())
+    {
+      defaultEndpointsProtocol = ite->second;
+    }
+
+    UrlBuilder builder;
+    builder.SetScheme(defaultEndpointsProtocol);
+    if (!blobEndpoint.empty())
+    {
+      builder = UrlBuilder(blobEndpoint);
+    }
+    else if (!accountName.empty())
+    {
+      builder.SetHost(accountName + ".blob" + defaultEndpointsProtocol);
+    }
+    else
+    {
+      throw std::runtime_error("invalid connection string");
+    }
+    builder.AppendPath(containerName, true);
+    builder.AppendPath(blobName, true);
+
+    auto credential = std::make_shared<SharedKeyCredential>(accountName, accountKey);
+
+    return BlobClient(builder.to_string(), credential, options);
+  }
+
+  BlobClient::BlobClient(
+      const std::string& blobUri,
+      std::shared_ptr<SharedKeyCredential> credential,
+      const BlobClientOptions& options)
+      : BlobClient(blobUri, options)
+  {
+    // not implemented yet
+    unused(credential, options);
+  }
+
+  BlobClient::BlobClient(
+      const std::string& blobUri,
+      std::shared_ptr<TokenCredential> credential,
+      const BlobClientOptions& options)
+      : BlobClient(blobUri, options)
+  {
+    // not implemented yet
+    unused(credential);
+  }
+
+  BlobClient::BlobClient(const std::string& blobUri, const BlobClientOptions& options)
+      : m_blobUrl(blobUri)
+  {
+    std::vector<std::unique_ptr<Azure::Core::Http::HttpPolicy>> policies;
+    for (const auto& p : options.policies)
+    {
+      policies.emplace_back(std::unique_ptr<Azure::Core::Http::HttpPolicy>(p->Clone()));
```

Comment:
Can we save the copy here by exposing another set of constructors?

Reply:
what the constructor would be like?